### PR TITLE
_versions file is now a separate chart

### PR DESCRIPTION
**Pull request by mvladev**
Rework the way charts are referenced:

`_versions` is now a new chart called `api-versions`. `terraformer-commons` is no longer archived, but symlinked.

If `api-versions` chart needs to be added to as a dependency to some other chart:

```bash
ln -sr ./charts/api-versions ./charts/PATH-TO-MY-CHART/charts

# for example

ln -sr ./charts/api-versions ./charts/seed-controlplane/charts/kube-apiserver/charts
```

**What this PR does / why we need it**: 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**: I've uncommented all `infa` charts - I don't think that it's a problem

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Can you elaborate on the problems?
Are you really sure that uncommenting the values in the infra charts is safe?
Have you tested the changes for all Kubernetes versions we are supporting?

**Comment by mvladev** [comment]
> Can you elaborate on the problems?

If you have symlink in one chart called `foo-chart` and then require this chart with `repository: file://../foo-chart` then in the resulting tar ball the symlinked file will be put in at path `Users/MY_USER/GO_PATH../github.com/gardener/gardener/chart/foo-chart/templates/SYMLINKED-FILE`

**Comment by rfranzke** [comment]
OK. Can you please rebase your branch? What about the remaining open questions from my previous comment?

**Comment by rfranzke** [comment]
ping

**Comment by mvladev** [comment]
@rfranzke I've all of the Infra charts have their default values overwritten in their respected  `.go` files.

**Comment by mvladev** [comment]
@rfranzke rebased to latest changes

**Comment by mvladev** [comment]
⚠️ #428 introduces a new symlink

**Comment by zanetworker** [comment]
@mvladev I ran some tests locally and it seems that the machine-controller-manager is crashlooping because of missing permissions  `machine-controller-manager 0/1       CrashLoopBackOff  `. Also found that the machine-controller-manager addon is not specified in the **requirements.yaml.**

The following is what I would expect:
```
- name: machine-controller-manager
  repository: file://addons/machine-controller-manager
  version: 0.1.0
```

Can you please run another test and confirm?

**Comment by rfranzke** [comment]
FYI: Due to @zanetworker's analysis (changes needed) I have now merged other PRs, e.g., #428 which you were pointing add, @mvladev. Please also rebase and adapt your changes! Thank you!

**Comment by rfranzke** [comment]
> Can you please run another test and confirm?

@mvladev, did you even test it yourself? I mean, when checking out the branch and creating a cluster it doesn't even come up. You would have seen this...
Instead, @zanetworker did now spend his time reviewing something that was not validated beforehand by the author itself...
Especially in our current situation (having very less output qualification) we should optimize this process. Please do at least at a base level test by creating a new cluster and updating an existing one.

**Comment by mvladev** [comment]
@rfranzke I've tested it, but I recall that the cluster was created with no problem.

The problem might be with reading in runtime of the `charts` folder. I often switch branches - I probably ran controller with my local `_versions` branch and switched to some other branch - the `charts` dir was checkout-ed and It rendered correctly.
As a workaround and to ensure stability and reliability, when the controller is started, we might want to copy the entire `charts` folder in a temporary dir and read charts from it.

Sorry for the wasted time @zanetworker.

**Comment by rfranzke** [comment]
Thanks @mvladev, no problem, I just wanted to raise the awareness.
@Adracus came up with https://github.com/kevinburke/go-bindata, what do you think about that?

**Comment by zanetworker** [comment]
@mvladev no worries :)

**Review by zanetworker — DISMISSED** [review]

**Comment by mvladev** [comment]
All issues should be fixed now - the cluster was created without any problems in 5ebf6b9 

The only problem seems to be that after every single `make helm-dep-update` all archived tars are updated. This is a new behavior - in the previous versions they remained unchanged.

**Comment by rfranzke** [comment]
OK, thanks. Running `make dep-helm-update` is only needed if the `api-versions` chart itself changes, right? Hence, shouldn't happen too often, I guess.

**Comment by mvladev** [comment]
@rfranzke it'll be needed also if you change one of the `shoot-addons` as well.

**Comment by rfranzke** [comment]
Oh ok, why?

File: charts/shoot-addons/requirements.yaml [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
   condition: kube2iam.enabled
 - name: kubernetes-dashboard
-  repository: http://localhost:10191
+  repository: file://addons/kubernetes-dashboard
+  version: 0.2.0
+  condition: kubernetes-dashboard.enabled
+- name: machine-controller-manager
+  repository: file://addons/machine-controller-manager
   version: 0.1.0
   condition: kubernetes-dashboard.enabled
```
Comment (rfranzke) [inline charts/shoot-addons/requirements.yaml]: copy-paste mistake?

**Review by rfranzke — CHANGES_REQUESTED** [review]
One small issue, otherwise looks good.
However, I really don't like that all the `tgz` files get modified on each and every single `make helm-dep-update` run. Can't this be worked around somehow? 😕

File: charts/shoot-addons/requirements.yaml (hunk 1 above)
Comment (mvladev) [reply charts/shoot-addons/requirements.yaml]: Yeah. This entire char should not be there in the first place. I've moved it to `shoot-core`

**Comment by mvladev** [comment]
> @rfranzke it'll be needed also if you change one of the shoot-addons as well.
> Oh ok, why?

Because, when using conditions in `requirements.yaml`, you can't reference them locally.:

```yaml
dependencies:
- name: cluster-autoscaler
  repository: file://charts/cluster-autoscaler
  version: 0.1.0
  condition: cluster-autoscaler.enabled
```

This returns an error. 

> However, I really don't like that all the tgz files get modified on each and every single make helm-dep-update run. Can't this be worked around somehow? 

Unfortunately, I'm not aware of way to do so.

**Comment by mvladev** [comment]
What we can do is not to check them in and when the controller-manager is started to build them dynamically

**Comment by rfranzke** [comment]
Hm yeah, but this would increase start-up time significantly, right? And, for local development purposes we need them to be checked in, don't we?

**Comment by mvladev** [comment]
What about -

1. We don't check them in git
1. When a developer clones the repo `make helm-dep-update` must be called.
1. The controller-manager can be started as before
1. On change in `charts` -> `make helm-dep-update`
1. On build for release -> `make helm dep-update`

**Comment by rfranzke** [comment]
Yes, we could do this. However, I especially don't like the fourth point. Can we get to a state where we ONLY need to run `make helm-dep-update` if we change the `api-versions` chart? I don't want to run it for each and every change. That requirement is highly error-prone, especially with local files ignored by git (the tgz's).

**Comment by rfranzke** [comment]
Short summary of a conversation with @mvladev: Martin will write a small program that gets called when using `make helm-dep-update`. It will be able to identify whether the Helm dependencies need to be updated or not, and will then only trigger the update for those requiring it.

**Comment by mvladev** [comment]
⚠️  PR has been updated with the new implementation. Checks for broken symlinks and template rendering have been added.

PR description has been changed to match with new changes

File: .ci/check [hunk 2]
```diff
@@ -46,3 +46,18 @@ go fmt ${PACKAGES}
...
+
+# Check for chart symlink errors
+BROKEN_SYMLINKS=$(find -L ${SOURCE_PATH}/charts -type l)
+if [[ -n ${BROKEN_SYMLINKS} ]]; then
+  echo "Found broken symlinks:"
+  echo ${BROKEN_SYMLINKS}
+  exit 1
+fi
+
+# Render chart dirs
```
Comment (rfranzke) [inline .ci/check:58]: Nice, thank you!

File: pkg/client/kubernetes/base/base.go [hunk 3]
```diff
@@ -57,11 +57,11 @@ func NewForConfig(config *rest.Config) (*Client, error) {
...
-			Pods:                   {"api", "v1"},
-			ReplicaSets:            {"apis", "apps", "v1"},
-			ReplicationControllers: {"api", "v1"},
-			Services:               {"api", "v1"},
-			StatefulSets:           {"apis", "apps", "v1"},
+			Pods:                      {"api", "v1"},
+			ReplicaSets:               {"apis", "apps", "v1"},
+			ReplicationControllers:    {"api", "v1"},
+			Services:                  {"api", "v1"},
+			StatefulSets:              {"apis", "apps", "v1"},
```
Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

File: .ci/check (hunk 2 above)
Comment (rfranzke) [inline .ci/check:58]: Nice, thank you!

File: pkg/client/kubernetes/base/base.go (hunk 3 above)
Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

File: pkg/utils/secrets/certificates.go [hunk 4]
```diff
@@ -174,11 +174,11 @@ func (s *CertificateSecretConfig) generateCertificateTemplate() *x509.Certificat
...
-			IsCA:         isCA,
-			SerialNumber: serialNumber,
-			NotBefore:    now,
-			NotAfter:     now.AddDate(10, 0, 0), // + 10 years
-			KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+			IsCA:                  isCA,
+			SerialNumber:          serialNumber,
+			NotBefore:             now,
+			NotAfter:              now.AddDate(10, 0, 0), // + 10 years
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
```
Comment (rfranzke) [inline pkg/utils/secrets/certificates.go:181]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

File: pkg/client/kubernetes/base/base.go (hunk 3 above)
Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor settings regarding auto-formatting, `make fmt/verify` shows formatting changes.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

File: .ci/check (hunk 2 above)
Comment (rfranzke) [inline .ci/check:58]: Nice, thank you!

Comment (rfranzke) [inline .ci/check:58]: Nice, thank you!

File: pkg/utils/secrets/certificates.go (hunk 4 above)
Comment (rfranzke) [inline pkg/utils/secrets/certificates.go:181]: You should fix your editor formatting settings.

File: pkg/client/kubernetes/base/base.go (hunk 3 above)
Comment (rfranzke) [inline pkg/client/kubernetes/base/base.go:64]: You should fix your editor formatting settings.

File: pkg/utils/secrets/certificates.go (hunk 4 above)
Comment (rfranzke) [inline pkg/utils/secrets/certificates.go:181]: You should fix your editor formatting settings.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
LGTM! Thank's for the check whether templates can be rendered!
PS: You should fix your editor formatting settings.

**Comment by rfranzke** [comment]
LGTM! Thank's for the check whether templates can be rendered!
PS: You should fix your editor formatting settings.

**Comment by rfranzke** [comment]
LGTM! Thank's for the check whether templates can be rendered!
PS: You should fix your editor formatting settings.

**Comment by mvladev** [comment]
This PR is merged, but for some reason github doesn't recognize it as such.
Closing it.

**Pull request merged**